### PR TITLE
Improvements to AnnotatedTypeMirror, fix for Issue 438, and misc. other fixes

### PR DESCRIPTION
**Pull request by wmdietl**

### CODE REVIEW

File: framework/src/org/checkerframework/framework/type/AnnotatedTypeMirror.java [hunk 1]
```diff
@@ -430,15 +430,29 @@ public boolean hasAnnotation(Class<? extends Annotation> a) {
     }
 
     /**
+     * Returns the actual effective annotation mirror used to annotate this type,
+     * whose Class equals the passed annoClass if one exists, null otherwise.
+     *
+     * @param annoClass annotation class
+     * @return the annotation mirror for anno
+     */
```
Comment (JonathanBurke) [inline framework/src/org/checkerframework/framework/type/AnnotatedTypeMirror.java]: Perhaps maybe add:
`@see getEffectiveAnnotations` for more information on effective annotations

File: framework/src/org/checkerframework/framework/type/AnnotatedTypeMirror.java [hunk 2]
```diff
@@ -1512,9 +1509,6 @@ void setLowerBoundField(AnnotatedTypeMirror type) {
          * @return the lower bound field.
          */
         public AnnotatedTypeMirror getLowerBoundField() {
-            if (lowerBound != null) {
```
Comment (JonathanBurke) [inline framework/src/org/checkerframework/framework/type/AnnotatedTypeMirror.java]: So I believe there might be a problem with this approach.  Currently, the annotations list in this class is protected which is bad.  It should be private.  So, if a class that has access to the annotations adds a primary annotation type variable through this mechanism, it won't add them to the bounds.  So if fixupBoundAnnotations is not called when getLowerBoundField/getLowerBound and getUpperBoundField/getUpperBound then the methods won't return the correct type.

Comment (JonathanBurke) [inline framework/src/org/checkerframework/framework/type/AnnotatedTypeMirror.java]: Note: I don't like this approach but it was a compromise I made while in the thick of type variable changes and on a deadline.  I guess the alternative would be to go through and fix all cases where annotations are accessed directly and put them through addAnnotation

File: framework/src/org/checkerframework/framework/type/AnnotatedTypeMirror.java (hunk 1 above)
Comment (wmdietl) [inline framework/src/org/checkerframework/framework/type/AnnotatedTypeMirror.java]: I agree that we should add more see annotations to javadoc, in particular in this class.
However, that seems like a separate, larger, clean-up that also explains the relationship between plain, effective, explicit, and all the other kinds of getter methods.

File: framework/src/org/checkerframework/framework/type/AnnotatedTypeMirror.java (hunk 2 above)
Comment (wmdietl) [inline framework/src/org/checkerframework/framework/type/AnnotatedTypeMirror.java]: Thanks for this comment!
I made the "annotations" field private and added a new protected getAnnotationsField method.
There are currently no calls of that method that modify the result. I documented that such modifications might break invariants.

I added a few other cleanups in the latest commits. Please merge if you like them.

**Pull request merged**